docs/proposals/parameterize-strategies.md

```diff
@@ -0,0 +1,267 @@
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+> 1. This will break the Build API.
+
+When should we do this? Do we need backwards compatibility?
+
+> 2. This will force strategy administrators to document required Parameters.
+
+How to do this?
+
+## Summary
+
+Today we use a feature of Tekton call [Parameters](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#specifying-parameters) when defining a `Task` spec. This feature allow us to populate a Parameter value with the users specifications from the `Build` API. In this way, we can manipulate on each strategy step arguments (_**arg** key_).
+
+## Motivation
+
+There are several reasons for the need of a more well define parameterization mechanism:
+
+- **Extensibility**: The current approach locks all strategies into three parameters, we need to allow N parameters if needed, depending on the strategy administrator requirement.
+
+- **API enhancement**: There are some fields in the Build API, like `spec.builder.image` that could belong to a parameterize block. Reserving a parent field in the Build API for a use-case that is not present across strategies is not ideal, it just generate complexity on the API level.
+
+- **Flexibility**: Providing a new API block on the `Build` or `BuildRun` for defining N parameters, will provide more flexibility for strategy administrators when defining new strategies. This will also simplify any potential documentation around "How to build a strategy?".
+
+- **Transparency**: At the moment is not clear which are the default parameters across strategies, for both strategy admins and Build users. Using Tekton Params nomenclature in the strategies will make it clearly.
+
+### Goals
+
+- Introduce a `spec.params` field across Build and BuildRun API´s, as a way for users to provide N parameters according to their preferred strategy of choice. Defining `spec.params` in a BuildRun overrides any definition of the same parameter in Build´s.
+
+- Introduce a `spec.params`on Strategies, both cluster or namespaced scope. This allow strategy administrators to layout the definition of N parameters, as required for their strategies.
+
+- Revisit what we consider today as default Parameters that are defined on runtime when creating `TaskRuns`. These are `DOCKERFILE`, `CONTEXT_DIR` and `BUILDER_IMAGE`. Decide if we need to reduce or extend the default definitions.
+
+- Expose the usage of Tekton Params directly on the strategies, in favor of simplicity and readability. This boils down to `inputs.params.DOCKERFILE`, instead of `build.dockerfile`.
+
+### Non-Goals
+
+- Build from scratch a paremeterize mechanism, by replacing the one available in Task/TaskRuns.
+
+- Provide support for Tekton Params of the type `array`.
```

Comment:
Can you elaborate on that use-case pls

Reply:
Build strategy:

```yaml
---
apiVersion: shipwright.io/v1alpha1
kind: ClusterBuildStrategy
metadata:
  name: kaniko
spec:
  params:
    - name: extraArgs
      description: allows to set arbitrary arguments for Kaniko
      type: array
  buildSteps:
    - name: build-and-push
      image: gcr.io/kaniko-project/executor:v1.5.2
      workingDir: /workspace/source
      command:
        - /kaniko/executor
      args:
        - --skip-tls-verify=true
        - --dockerfile=$(build.dockerfile)
        - --context=/workspace/source/$(build.source.contextDir)
        - --destination=$(build.output.image)
        - --oci-layout-path=/workspace/output/image
        - --snapshotMode=redo
        - --push-retry=3
        - $(params.extraArgs[*])
```

Build:

```yaml
---
apiVersion: shipwright.io/v1alpha1
kind: Build
metadata:
  name: kaniko-golang-build
  annotations:
    build.shipwright.io/build-run-deletion: "true"
spec:
  source:
  strategy:
    name: kaniko
    kind: ClusterBuildStrategy
  output:
  params:
    - name: extraArgs
      values:
        - --buildArg
        - KEY=VALUE
        - --buildArg
        - ANOTHERKEY=anotherValue
```

That's actually not ideal because then a user can set arbitrary Kaniko arguments, potentially those that the build strategy administrator does not want. So, it needs some trust from the Build user. What I am instead looking for is this in the build strategy:

```yaml
  params:
    - name: buildArgs
      description: allows to set the build arguments similar to https://docs.docker.com/engine/reference/commandline/build/#set-build-time-variables---build-arg; values need to be in the form KEY=VALUE
      type: array
```

But this, I cannot express in the steps without using a script that takes the values from `buildArgs` and somehow adds the `--buildArg` argument to it, or something like the following which I consider over-engineered for the start:

```yaml
  params:
    - name: buildArgs
      description: allows to set the build arguments similar to https://docs.docker.com/engine/reference/commandline/build/#set-build-time-variables---build-arg; values need to be in the form KEY=VALUE
      type: array
      mapperFunction: "(value) => ['--build-arg', value]"
```

So, is fine if you start with single values only. We can add array support later.

Reply:
I also agree we don't need to forbid the array type. If Tekton supports it natively, we just pass what admin/user defines. I don't have a good example now, but such as the linux capability?: https://github.com/shipwright-io/build/blob/master/samples/buildstrategy/source-to-image/buildstrategy_source-to-image_cr.yaml#L40-L46

Reply:
Thanks, there is definitely a value on having that but also a risk on providing too much freedom to users. Based on both of your comments, I´m enhancing the sentence as follows:

> Provide support for Tekton Params of the type `array`. This does not mean we block support for this feature in the future.